internal/provisioner/helm_utils.go

```diff
@@ -0,0 +1,226 @@
+package provisioner
+
+import (
+	"bytes"
+	"context"
+	"os/exec"
+	"time"
+
+	"github.com/mattermost/mattermost-cloud/internal/tools/k8s"
+	"github.com/mattermost/mattermost-cloud/internal/tools/kops"
+	"github.com/mattermost/mattermost-cloud/model"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+// helmDeployment deploys Helm charts.
+type helmDeployment struct {
+	chartDeploymentName string
+	chartName           string
+	namespace           string
+	setArgument         string
+	valuesPath          string
+
+	cluster         *model.Cluster
+	kopsProvisioner *KopsProvisioner
+	kops            *kops.Cmd
+	logger          log.FieldLogger
+}
+
+func installHelm(kops *kops.Cmd, logger log.FieldLogger) error {
+	logger.Info("Installing Helm")
+
+	err := helmSetup(logger, kops)
+	if err != nil {
+		return err
+	}
+
+	wait := 120
+	logger.Infof("Waiting up to %d seconds for helm to become ready...", wait)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(wait)*time.Second)
+	defer cancel()
+	err = waitForHelmRunning(ctx, kops.GetKubeConfigPath())
+	if err != nil {
+		return err
+	}
+
+	logger.Info("Updating all Helm repos.")
+	return helmRepoUpdate(logger)
+}
+
+func (d *helmDeployment) Update() error {
+	logger := d.logger.WithField("helm-update", d.chartName)
+
+	logger.Infof("Refreshing helm chart %s -- may trigger service upgrade", d.chartName)
+	err := upgradeHelmChart(*d, d.kops.GetKubeConfigPath(), logger)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Create will install a given HelmDeployment in the cluster specified in that object.
+func (d *helmDeployment) Create() error {
+	logger := d.logger.WithField("helm-create", d.chartName)
+	logger.Infof("Installing helm chart %s", d.chartName)
+	return installHelmChart(*d, d.kops.GetKubeConfigPath(), logger)
+}
+
+// waitForHelmRunning is used to check when Helm is ready to install charts.
+func waitForHelmRunning(ctx context.Context, configPath string) error {
+	for {
+		cmd := exec.Command("helm", "ls", "--kubeconfig", configPath)
+		var out bytes.Buffer
+		cmd.Stderr = &out
+		cmd.Run()
+		if out.String() == "" {
+			return nil
+		}
+		select {
+		case <-ctx.Done():
+			return errors.Wrap(ctx.Err(), "timed out waiting for helm to become ready")
+		case <-time.After(5 * time.Second):
+		}
+	}
+}
+
+// helmRepoUpdate updates the helm repos to get latest available charts
+func helmRepoUpdate(logger log.FieldLogger) error {
+	var cmd *exec.Cmd
+	arguments := []string{
+		"repo",
+		"update",
+	}
+
+	cmd = exec.Command("helm", arguments...)
+
+	logger.WithFields(log.Fields{
+		"cmd": cmd.Path,
+	}).Info("Invoking command")
+
+	err := cmd.Run()
+	return err
```

Comment:
nit: you have a few different ways of returning errors across your functions mismatch across your functions (you sometimes check the error and sometimes not). It would be nice if you standardized on one.

Reply:
Ah, yeah this is a style thing, what you're seeing is that it's redundant to do 
```go
if err != nil {
   return err
}
return nil
```
if the function producing `err` is the last operation in the routine and the routine only returns `error` type. I can definitely change these to be more redundant but also more explicit if that's the project style!

Reply:
Though the quoted one is silly, I assign `err` and then return, lol, so I'll go through and make sure I'm consistent.

Reply:
Yeah I see what you're saying, I will standardize these error returns (and also not do something silly like assigning `err` and then just returning it lol)

I'm also just going to leave this note to myself to go back and add a little bit of debugging info instead of just returning `err` as-is in most of these cases (using `errors.Wrap()`) now that the functionality is nailed down.